### PR TITLE
Trigger discovery on interpreter/package changes

### LINKED ISSUE #5424: Pytest: should trigger test discovery when interpreter factory changed or installed packages changed
If you open a project that has pytest enabled but no pytest installed, then install pytest in your environment, the tests are not discovered once pytest has installed.

If you change the environment between env that have pytest installed and others that don't, discovery should be triggered.

So we should trigger discovery when:
- current interpreter has changed (listen for that event)
- installed packages in current interpreter changed (IPackageManager.InstalledPackagesChanged)

Note that the interpreter can change without necessarily changing any project property. Example, a project uses the default global interpreter, and the user changes the global default to be a different interpreter.

### PR DESCRIPTION
**Pull request by huguesv**
Fix #5424

See commit messages (it may be helpful to look at them individually).

### CODE REVIEW

File: Python/Product/TestAdapter/TestContainerDiscovererProject.cs [hunk 1]
```diff
@@ -274,6 +288,7 @@ class TestContainerDiscovererProject : ITestContainerDiscoverer, IDisposable {
             var pyProj = PythonProject.FromObject(e.Project);
             if (pyProj != null) {
                 pyProj.ProjectPropertyChanged -= OnTestPropertiesChanged;
+                pyProj.ActiveInterpreterChanged -= OnActiveInterpreterChanged;
```
Comment (bschnurr) [inline Python/Product/TestAdapter/TestContainerDiscovererProject.cs]: do we need to _packageManagerEventSink.Unwatch(project)  here?

File: Python/Product/VSInterpreters/Interpreter/PythonWorkspaceContext.cs [hunk 2]
```diff
@@ -265,15 +265,16 @@ class PythonWorkspaceContext : IPythonWorkspaceContext {
...
                 var oldSearchPaths = _searchPaths;
-                searchPathsChanged = !oldSearchPaths.SequenceEqual(_searchPaths);
                 _searchPaths = ReadSearchPathsSetting();
 
                 var oldTestFramework = _testFramework;
                 _testFramework = GetStringProperty(TestFrameworkProperty);
+
+                interpreterChanged = oldInterpreter != _interpreter;
+                searchPathsChanged = !oldSearchPaths.SequenceEqual(_searchPaths);
                 testSettingsChanged = !String.Equals(oldTestFramework, _testFramework);
```
Comment (bschnurr) [inline Python/Product/VSInterpreters/Interpreter/PythonWorkspaceContext.cs]: good catch.

File: Python/Product/VSInterpreters/PackageManager/CPythonPipPackageManagerProvider.cs [hunk 3]
```diff
@@ -35,27 +35,42 @@ class PipCommandsV27AndLater : PipPackageManagerCommands {
...
+                    }
+                }
+            }
+
+            if (pm != null) {
+                yield return pm;
+            }
+        }
+
+        private IPackageManager TryCreatePackageManager(IPythonInterpreterFactory factory) {
```
Comment (bschnurr) [inline Python/Product/VSInterpreters/PackageManager/CPythonPipPackageManagerProvider.cs]: best to check for factory being null

File: Python/Product/TestAdapter/TestContainerDiscovererProject.cs (hunk 1 above)
Comment (huguesv) [reply Python/Product/TestAdapter/TestContainerDiscovererProject.cs]: I don't think so. When any change requires refresh, we clear all. Then every project will add its interpreter factory package managers.

Comment (bschnurr) [reply Python/Product/TestAdapter/TestContainerDiscovererProject.cs]: ahh..  I need to look again if calling refresh is really necessary when a project unloads. for now this works

**Pull request merged**